xdelta3-dir-patcher

```diff
@@ -137,29 +546,53 @@ class XDelta3DirPatcher(object):
 
             self.copy_attributes(patch_path, target_path)
 
+            remove(patch_path)
+
     # TODO: Unit test me
-    def diff(self, old_dir, new_dir, patch_bundle, metadata = None):
-        # Create a temp dir
-        target_dir = mkdtemp(prefix="%s_" % self.__class__.__name__)
+    def diff(self, old_dir, new_dir, patch_bundle, metadata = None,
+             staging_dir = None,
+             runner = ExecutorRunner()):
+        target_dir = mkdtemp(prefix='%s_target' % XDelta3DirPatcher.__name__,
+                             dir=staging_dir)
+
         print("Using %s as staging area" % target_dir)
+        stdout.flush()
 
         delta_target_dir = path.join(target_dir, self.PATCH_FOLDER)
-        mkdir(delta_target_dir)
+        if not path.isdir(delta_target_dir):
+            mkdir(delta_target_dir)
+
+        with XDeltaArchive(old_dir) as old_archive_obj, \
+             XDeltaArchive(new_dir) as new_archive_obj:
+            old_staging_dir = mkdtemp(prefix='%s_old_src' % XDelta3DirPatcher.__name__,
+                                      dir=staging_dir)
+            new_staging_dir = mkdtemp(prefix='%s_new_src' % XDelta3DirPatcher.__name__,
+                                      dir=staging_dir)
 
-        for root, new_dirs, new_files in walk(new_dir):
-            print("Entering directory", root)
+            # XXX: Preload old archive file list to avoid concurrency issues
+            # TODO: Make the listing thread-safe
+            old_archive_obj.list_files()
 
-            rel_path = path.relpath(root, new_dir)
+            for filename in new_archive_obj.list_files():
+                if self.args.debug:
+                    print('Processing %s' % filename)
+                else:
+                    print('.', end = "")
 
-            if self.args.debug: print("- Subdirs:", new_dirs)
-            if self.args.debug: print("- Files:", new_files)
+                runner.add_task(self._find_file_delta, (filename,
+                                                        old_archive_obj,
+                                                        new_archive_obj,
+                                                        old_staging_dir,
+                                                        new_staging_dir,
+                                                        delta_target_dir))
```

Comment:
Tracking dependencies could be tricky. That's why I was thinking a 2 pass approach be easiest. Extract and diff all the files, simply doing a regular `makedirs` to create directories where necessary. Then go through the directories, doing `makedirs` to ensure they exist then setting the ownership/permissions. Hmm, that still could suffer from the same issue where you wouldn't be able to change attributes of a subdir once the parent dir becomes read only.